### ISSUE TITLE
Progress Meter

### ISSUE THREAD

**Issue by Thom1729** [SublimeText/sublime_lib#114]
Many packages display a progress meter in the status bar. Although there is no standard implementation, most packages have seemed to converge upon the same format: `<message> [ =    ]`. An implementation of this common pattern would be handy for package developers, especially those who might not otherwise be motivated to write the async code to make it work.

The Package Control progress bar is probably a good model (for UI, if not necessarily for implementation). Desirable features:

- Initialize with a window (call `window.status_message`), with a view and an optional key (call `view.set_status()`/`view.erase_status()`), or with nothing (call `sublime.status_message`).
- Work as a context manager (`with ProgressMeter(...):`) or manually (`progress_meter.start()`).
- Use a dedicated thread and `sleep` rather than the shared threads.
- Lock around start/stop for thread safety.
- Provide an overrideable `get_display(iteration)` method for subclassing.
- Optional: specify an automatic termination condition.
- Optional: specify success and failure status messages.

Thoughts?

**Comment by FichteFoll** [comment]
Considering how often this is used by packages, I believe it is sensible to bundle this in sublime_lib. Although the more common concept is a "loading bar" indicating that something is happening rather than how far into a progress we are. Iirc Microsoft uses the term "marquee" for this type of progress bar.

Anyway, I remember stubling across a repo that hat various loading bar profiles to switch between (with each state being represented as a string) as well as a default timeout between states. I'd like to adopt an API that allows custom loading bars like this but with a reasonable default implementation, e.g. Package Control's. However, we can of course still implement this in an overridable method for even more flexibiltiy.
The only other API we need is to start and stop the bar, which can easily be a direct subclass of a thread with a stop method and the context manager protocol. For state tracking, use an `Event`.
As you suggested, I believe a custom thread is most appropriate as we cannot rely on the async thread not being blocked.

For automatic termination I suppose you're thinking of a timeout? I don't think we realistically need that, do we? If users use the context manager, it will be cleaned up regardless as soon as the block is left. If they invoke it manually, they better keep around their thread reference and look to have a timeout for their process (whatever they are doing) themselves. Providing a timeout function would push anti-patterns.

What would be the advantage over specifying success and failure status messages compared to using `sublime.status_message` directly?

**Comment by Thom1729** [comment]
> Although the more common concept is a "loading bar" indicating that something is happening rather than how far into a progress we are. Iirc Microsoft uses the term "marquee" for this type of progress bar.

Agreed. I see it called a "progress bar" all the time, but it's not really showing progress. The most general term might be "loading indicator" (making the fewest assumptions about presentation), but I'm not attached to anything specific.

> For automatic termination I suppose you're thinking of a timeout? I don't think we realistically need that, do we? If users use the context manager, it will be cleaned up regardless as soon as the block is left. If they invoke it manually, they better keep around their thread reference and look to have a timeout for their process (whatever they are doing) themselves. Providing a timeout function would push anti-patterns.

Agreed. If the indicator isn't stopped, that's programmer error, and offering a fallback for that is just encouraging bad behavior.

I was also looking at Package Control's `ThreadProgress` class, which polls `self.thread.is_alive()`. This seems like a reasonable use case:

```py
target = partial(install_package, self.manager, name, on_complete)
thread = Thread(target=target)
thread.start()

ProgressBar(
    'Installing package %s' % name,
    success_message='Package %s successfully %s' % (name, self.completion_type),
    run_while=thread.is_alive
).start()
```

Alternatively, that particular use case could be supported by making `ProgressBar` a `ContextDecorator`:

```py
progress = ProgressBar(
    'Installing package %s' % name,
    success_message='Package %s successfully %s' % (name, self.completion_type),
)
target = partial(install_package, self.manager, name, on_complete)
thread = Thread(target=progress(target))
thread.start()
```

Now that I've gone through it, that's probably better anyway.

> What would be the advantage over specifying success and failure status messages compared to using `sublime.status_message` directly?

Example from UnitTesting:

```py
def reload_package(self, package, dummy=False, show_reload_progress=False):
    if show_reload_progress:
        progress_bar = ProgressBar("Reloading %s" % package)
        progress_bar.start()

        try:
            reload_package(package, dummy=dummy, verbose=True)
        except Exception:
            sublime.status_message("Fail to reload {}.".format(package))
        finally:
            progress_bar.stop()

        sublime.status_message("{} reloaded.".format(package))
    else:
        reload_package(package, dummy=dummy, verbose=False)
```

I think that this is a pretty common pattern: showing a message after the indicator is finished, and optionally a different message in case of error. In this case, the `except` statement hides the error, which is probably undesirable. If `ProgressBar` is a context manager, we can simplify it thus:

```py
def reload_package(self, package, dummy=False, show_reload_progress=False):
    if show_reload_progress:
        with ProgressBar("Reloading %s" % package):
            try:
                reload_package(package, dummy=dummy, verbose=True)
            except Exception:
                sublime.status_message("Fail to reload {}.".format(package))
                raise
            else:
                sublime.status_message("{} reloaded.".format(package))
    else:
        reload_package(package, dummy=dummy, verbose=False)
```

This still duplicates the same primary code (calling `reload_package`). We can refactor this (using e.g. `contextlib.ExitStack`):

```py
def reload_package(self, package, dummy=False, show_reload_progress=False):
    try:
        with ExitStack() as stack:
            if show_reload_progress:
                stack.enter_context(ProgressBar("Reloading %s" % package))

            reload_package(package, dummy=dummy, verbose=True)

    except Exception:
        if show_reload_progress:
            sublime.status_message("Fail to reload {}.".format(package))
        raise
    else:
        if show_reload_progress:
            sublime.status_message("{} reloaded.".format(package))
```

The repeated `if` statements are a bit of a mess. Also, this doesn't work -- `ProgressBar.__exit__()` will clobber the messages. We need to reverse the nesting:

```py
def reload_package(self, package, dummy=False, show_reload_progress=False):
    try:
        with ExitStack() as stack:
            if show_reload_progress:
                stack.enter_context(ProgressBar("Reloading %s" % package))

            reload_package(package, dummy=dummy, verbose=True)

    except Exception:
        if show_reload_progress:
            sublime.status_message("Fail to reload {}.".format(package))
        raise
    else:
        if show_reload_progress:
            sublime.status_message("{} reloaded.".format(package))
```

If `ProgressBar` handles success and error messages, this becomes much simpler:

```py
def reload_package(self, package, dummy=False, show_reload_progress=False):
    with ExitStack() as stack:
        if show_reload_progress:
            stack.enter_context(ProgressBar(
                "Reloading %s" % package,
                success_message="{} reloaded.".format(package),
                error_message="Fail to reload {}.".format(package),
            ))

        reload_package(package, dummy=dummy, verbose=True)
```

At least in this example, the success/error status messages seem to logically be part of the loading indicator.

**Comment by FichteFoll** [comment]
>loading indicator
:+1: 

>Alternatively, that particular use case could be supported by making `ProgressBar` a `ContextDecorator`:
I like decorators (a lot), so I prefer this. The only issue I'm seeing is if we require a window/view parameter but don't have one, though we can resort to resolving this through `sublime.active_{window,view}()`.

Your ExitStack example with the optional loading indicator seems a bit specific to me. Why does `ProgressBar.__exit__()` clobber the messages, though? Shouldn't it be called before the exception is caught by the surrounding try-except?
I can see that the code paths are much clearer in the final example, but wouldn't a designated class be clearer? You're using `ExitStack` anyway and I don't see a much simpler approach.

```py
class StatusReporter:
    def __init__(self, success_message=None, error_message=None):
        self.success_message = success_message
        self.error_message = error_message

    def __enter__(self):
        return self

    def __exit__(self, *args):
        if args[0]:
            sublime.status_message(self.error_message)
        else:
            sublime.status_message(self.success_message)
        

def reload_package(self, package, dummy=False, show_reload_progress=False):
    with ExitStack() as stack:
        if show_reload_progress:
            stack.enter_context(StatusReporter(
                success_message="{} reloaded.".format(package),
                error_message="Fail to reload {}.".format(package),
            ))
            stack.enter_context(ProgressBar("Reloading %s" % package))

        reload_package(package, dummy=dummy, verbose=True)
```

**Comment by Thom1729** [comment]
> The only issue I'm seeing is if we require a window/view parameter but don't have one, though we can resort to resolving this through `sublime.active_{window,view}()`

We can default to `sublime.status_message()`, which seems to work on the active window, though it's undocumented. I think that requiring an argument is reasonable; the caller can always provide `sublime.active_window()` if that makes sense.

> loading indicator

Second-guessing myself, maybe "loading" is even too specific. Maybe a "wait" indicator?

> Why does `ProgressBar.__exit__()` clobber the messages, though? Shouldn't it be called before the exception is caught by the surrounding try-except?

I posted the wrong snippet by mistake; it's the same as the following snippet that works correctly. The intended, faulty snippet is:

```py
def reload_package(self, package, dummy=False, show_reload_progress=False):
    with ExitStack() as stack:
        if show_reload_progress:
            stack.enter_context(ProgressBar("Reloading %s" % package))
        try:
            reload_package(package, dummy=dummy, verbose=True)

        except Exception:
            if show_reload_progress:
                sublime.status_message("Fail to reload {}.".format(package))
            raise
        else:
            if show_reload_progress:
                sublime.status_message("{} reloaded.".format(package))
```

It looked like a possible footgun to me.

> I can see that the code paths are much clearer in the final example, but wouldn't a designated class be even clearer? You're using `ExitStack` already and I don't see a much simpler approach anyway.

The order of the context managers is significant. If they are interchanged, then the code will run but the result will be wrong. If we provide both a `LoadingIndicator` and a `StatusReporter`, given that we anticipate that they would often be used together, then it seems likely that some users would put them in the wrong order. This is why it seems like a possible footgun.

In principle, this is because both context managers are trying to own the same piece of state, and the underlying Sublime API doesn't have any mechanism for making this safe. We could partially mitigate this by adding some internal checks to display a warning if two indicators are trying to use the same "messageable" (window or (view, key) pair) at the same time (using a global `WeakKeyDictionary` and some locking). The `ProgressBar` would claim ownership when it starts and give it up when it stops, and the `StatusReporter` would claim it only briefly while printing messages. This way, if the context managers were in the wrong order, then there would at least be a warning printed.

**Comment by Thom1729** [comment]
Closed by #119.
